crypt-gost-yescrypt.c

```diff
@@ -0,0 +1,364 @@
+
+/* HMAC_GOSTR3411_2012_256 */
+static void
+gost_hmac256(const uint8_t *k, size_t n, const uint8_t *t, size_t len,
+             uint8_t *out32, crypt_gost_hmac_256_t *gostbuf)
+{
+  size_t i;
+
+  /* R 50.1.113-2016 only allowed N to be in range 256..512 bits */
+  assert(n >= GOSTR3411_2012_IN && n <= GOSTR3411_2012_OUT);
+
+  for (i = 0; i < sizeof(gostbuf->pad); i++)
+    gostbuf->kstar[i] = i < n ? k[i] : 0;
+
+  GOST34112012Init (&gostbuf->ctx, GOSTR3411_2012_BITS);
+  for (i = 0; i < sizeof(gostbuf->pad); i++)
+    gostbuf->pad[i] = gostbuf->kstar[i] ^ 0x36; /* ipad */
+  GOST34112012Update (&gostbuf->ctx, gostbuf->pad,
+                      sizeof(gostbuf->pad));
+  GOST34112012Update (&gostbuf->ctx, t, len);
+  GOST34112012Final (&gostbuf->ctx, gostbuf->digest);
+
+  GOST34112012Init (&gostbuf->ctx, GOSTR3411_2012_BITS);
+  for (i = 0; i < sizeof(gostbuf->pad); i++)
+    gostbuf->pad[i] = gostbuf->kstar[i] ^ 0x5c; /* opad */
+  GOST34112012Update (&gostbuf->ctx, gostbuf->pad,
+                      sizeof(gostbuf->pad));
+  GOST34112012Update (&gostbuf->ctx, gostbuf->digest,
+                      sizeof(gostbuf->digest));
+  GOST34112012Final (&gostbuf->ctx, out32);
+}
+
+/*
+ * As OUTPUT is initialized with a failure token before gensalt_yescrypt_rn
+ * is called, in case of an error we could just set an appropriate errno
+ * and return.
+ */
+void
+gensalt_gost_yescrypt_rn (unsigned long count,
+                          const uint8_t *rbytes, size_t nrbytes,
+                          uint8_t *output, size_t o_size)
+{
+  uint8_t *tmp = malloc (o_size);
+
+  /* Check for failures.  */
+  if (tmp == NULL)
+    return;
+
+  /* Initialize stack buffer with the failure token from output.  */
+  XCRYPT_STRCPY_OR_ABORT (tmp, o_size, output);
+
+  /* We pass 'o_size - 1' to gensalt, because we need to expand
+           the prefix by 1 char to insert the gost marker.  */
+  gensalt_yescrypt_rn (count, rbytes, nrbytes, tmp, o_size - 1);
+
+  /* Check for failures.  */
+  if (*tmp == '*')
+    return;
+
+  /* Copy to output and clean the stack.  */
+  XCRYPT_STRCPY_OR_ABORT (output + 2, o_size - 1, tmp + 1);
+  XCRYPT_SECURE_MEMSET (tmp, CRYPT_GENSALT_OUTPUT_SIZE);
+
+  /* Prepend with the gost marker.  */
+  output[0] = '$';
+  output[1] = 'g';
+}
+
+void
+crypt_gost_yescrypt_rn (const char *phrase, size_t phr_size,
+                        const char *setting, size_t set_size,
+                        uint8_t *output, size_t o_size,
+                        void *scratch, size_t s_size)
+{
+  if (o_size < set_size + 1 + 43 + 1 ||
+      CRYPT_OUTPUT_SIZE < set_size + 1 + 43 + 1 ||
+      s_size < sizeof (crypt_gost_yescrypt_internal_t))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  /* Fail when called with wrong prefix.  */
+  if (strncmp (setting, "$gy$", 4))
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  crypt_gost_yescrypt_internal_t *intbuf = scratch;
+
+  if (yescrypt_init_local (&intbuf->local))
+    return;
```

Comment:
I see no point in keeping `local` inside of `stratch` if we re-initialize it every time anyway.

Reply:
I did it the same way for `crypt-yescrypt`.  The scratch buffer is big enough and initialized on every call to `crypt()`, thus if we put `local` there, we save some space on the stack.